packages/bot/src/stores/run-panel-store.js

```diff
@@ -68,10 +70,26 @@ export default class RunPanelStore {
             return;
         }
 
-        this.registerBotListeners();
+        const top_blocks = Blockly.derivWorkspace.getTopBlocks();
+        const blocks_in_workspace = Blockly.derivWorkspace.getAllBlocks();
+        const has_tradeoptions = Object.keys(blocks_in_workspace)
+            .some(key => blocks_in_workspace[key].type === 'trade_definition_tradeoptions');
+
+        const { mandatoryMainBlocks } = config;
+        const has_mandatory_blocks = mandatoryMainBlocks
+            .every(block_type => top_blocks.some(top_block => top_block.type === block_type));
+
+        if (!has_mandatory_blocks || !has_tradeoptions) {
```

Comment:
I think it would be clearer to create an array of required block types and then check whether all of these blocks could be found in the workspace. Example:

```JavaScript
    const required_block_types    = ['trade_definition_tradeoptions', ...config.mandatoryMainBlocks];
    const all_block_types         = Blockly.derivWorkspace.getAllBlocks().map(block => block.type);
    const has_all_required_blocks = required_block_types.every(required_block_type => all_blocks.includes(required_block_type));

    if (!has_all_required_blocks) {
      // Logic.
    }
```